core/sys/pt.h

```diff
@@ -296,6 +296,34 @@ struct pt {
   } while(0)
 
 /**
+ * \brief Yield from the current protothread then execute a statement
+ *
+ * This function will yield the protothread, thereby allowing other
+ * processing to take place in the system.
+ *
+ * This macro is intended to be used when there is a possibility that the
+ * given statement will cause recursion into the protothread either directly
+ * or through an ISR. Without this construction, the protothread would not
+ * correctly continue from the yield point.
+ *
+ * Note that proto-threads are never truly reentrent.
+ *
+ * \param pt A pointer to the protothread control structure.
+ * \param stmnt code statement to execute before yielding
+ *
+ * \hideinitializer
+ */
+#define PT_YIELD_ON(pt,stmnt)                         \
+  do{                                                 \
+    PT_YIELD_FLAG = 0;                                \
+    LC_SET((pt)->lc);                                 \
+    if(PT_YIELD_FLAG == 0) {                          \
+      { stmnt ;}                                      \
+      return PT_YIELDED;                              \
+    }                                                 \
+  }while(0)
```

Comment:
The difference is when the "stmnt" causes re-entry back into the proto-thread. With PT_YIELD_ON the protothread will continue execution after the macro on re-entry and "stmnt" will not be run twice.  If we instead execute "stmnt" before a regular PT_YIElD then on re-entry the protothread starts execution at the point where it was previously suspended and stmnt executes again.

> On Mar 11, 2016, at 6:01 AM, greg-king5 <notifications@github.com> wrote:
> 
> I'm confused. How is that macro better than writing simply:
> 
> stmnt;
> PT_YIELD(pt);
> Why should the code have lines such as
> 
> PT_YIELD_ON(&cca_state.pt, call_off(cmd_cb_cca_sm));
> instead of
> 
> call_off(cmd_cb_cca_sm);
> PT_YIELD(&cca_state.pt);
> I don't see any significant difference between them -- even if stmnt; is recursive.
> 
> In other words, why should that macro be added to Contiki? (Its name implies something that isn't true; there is no conditional relationship between the statement and the yield.)
> 
> —
> Reply to this email directly or view it on GitHub.
>